### PR TITLE
Adding `ModalFooter` component

### LINKED ISSUE #18034: Create component: ModalFooter
### Description
Create a reusable UI component for `ModalFooter` according to the insight report

### References
[Modal Audit FigJam](https://www.figma.com/file/hxYqloYgmVcgsoiVqmGZ8K/Modal-Audit?node-id=0%3A1&t=WcWEbEnjEkGuOdXr-1)
[Figma component](https://www.figma.com/file/HKpPKij9V3TpsyMV1TpV7C/DS-Components?node-id=3180%3A13642)
[UI component guidelines](https://www.notion.so/UI-Components-5ccd2bf83eb8441892a0b72c0d8929e1)
[Styling guidelines](https://www.notion.so/Extension-Frontend-Engineering-Guide-b82ddb3e14004b5db3799a9b446294a9#49aa2ef7b7aa40178ff918043df71c34)
Testing guidelines: TBC

### Files needed
- `index.js`
- `index.scss`
- `modal-footer.js`
- `modal-footer.stories.js`
- `modal-footer.test.js`
- `README.mdx`

### Technical details
TBC

### Acceptance criteria
- [ ] Uses color, typography, shadows design tokens 
- [ ] Uses semantic html
- [ ] PropTypes have descriptions
- [ ] Has storybook story with controls
- [ ] Has documentation in MDX
- [ ] Has unit tests and 90% coverage
- [ ] Works in Chrome and Firefox
- [ ] Performance tested: no unnecessary re-renders or other performance concerns

### PR DESCRIPTION
**Pull request by georgewrmarshall**
## **Description**

Creating the `ModalFooter` component to provide a consistent consolidated footer for all Modal UI

## **Related issues**

Fixes: #18034

## **Manual testing steps**

1. Go to the storybook build of this PR
2. Search `ModalFooter` in the search bar
3. Find `ModalFooter` component
4. Read documentation, check stories, check controls

### **Before**

No `ModalFooter` component existed before. I have added before after to other modal sub components in comments

### **After**

https://github.com/MetaMask/metamask-extension/assets/8112138/bc2067b6-26f5-4576-97ec-43e8f9389af5

100% test coverage

Screenshot 2024-01-04 at 6 08 57 PM

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/component-library/button/index.ts [hunk 1]
```diff
@@ -14,7 +15,7 @@ export const BUTTON_SIZES = {
   AUTO: Size.auto,
 };
 /**
- * @deprecated `BUTTON_SIZES` const has been deprecated in favor of the `ButtonSize` enum which can still be imported from `ui/components/component-library`
+ * @deprecated `BUTTON_VARIANT` const has been deprecated in favor of the `ButtonVariant` enum which can still be imported from `ui/components/component-library`
  */
 export const BUTTON_VARIANT = {
   PRIMARY: 'primary',
```
Comment (georgewrmarshall) [inline ui/components/component-library/button/index.ts:21]: Fixing incorrect deprecation messages above old constants

File: ui/components/component-library/container/README.mdx [hunk 2]
```diff
@@ -19,7 +19,7 @@ Using the `maxWidth` prop with `ContainerMaxWidth` enum values, you can set the
 
 Possible values are:
 
-- `ContainerMaxWidth.Sm` (max-width: 320px)
+- `ContainerMaxWidth.Sm` (max-width: 360px)
```
Comment (georgewrmarshall) [inline ui/components/component-library/container/README.mdx:22]: Fixing incorrect container widths

File: ui/components/component-library/container/index.ts [hunk 3]
```diff
@@ -1,2 +1,3 @@
 export { Container } from './container';
 export type { ContainerProps } from './container.types';
+export { ContainerMaxWidth } from './container.types';
```
Comment (georgewrmarshall) [inline ui/components/component-library/container/index.ts:3]: Adding missing enum export

File: ui/components/component-library/modal-content/modal-content.stories.tsx [hunk 4]
```diff
@@ -185,11 +179,10 @@ export const Size: StoryFn<typeof ModalContent> = (args) => {
...
-                Close
-              </Button>
-            </Box>
+            <ModalFooter
+              onSubmit={() => setCurrentSize(null)}
+              submitButtonProps={{ children: 'Close' }}
+            />
           </ModalContent>
         </Modal>
       )}
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-content/modal-content.stories.tsx:188]: Replacing placeholders in `ModalContent` stories and docs with `ModalFooter`

### Before

https://github.com/MetaMask/metamask-extension/assets/8112138/ada03463-ce54-4995-9cff-80fa29f802bd



### After


https://github.com/MetaMask/metamask-extension/assets/8112138/3b4c68d5-f2da-4618-ad5e-eb63df628f6d

File: ui/components/component-library/modal-footer/README.mdx [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  onCancel={handleOnCancel}
+  cancelButtonProps={{
+    children: 'Cancel this',
+  }}
+/>;
+```
+
+### Children
+
+Use the `children` prop to pass any in any valid React children to the `ModalFooter`. The children will appear above the submit and cancel buttons and outside of the internal `Container` component. If you require children above the buttons, you can build your own `ModalFooter` component using the `Button` and `Container` component.
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-footer/README.mdx:62]:
```suggestion
Use the `children` prop to pass any in any valid React children to the `ModalFooter`. The children will appear above the submit and cancel buttons and outside of the internal `Container` component. If you require children above the buttons, you can build your own `ModalFooter` component using the `Button` and `Container` components.
```

File: ui/components/component-library/modal-footer/README.mdx [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+      isChecked={checked}
+      onChange={handleCheckboxChange}
+    />
+  </Container>
+</ModalFooter>;
+```
+
+### Container Props
+
+The `ModalFooter` has an internal container that prevents the buttons from being large widths. You can override these props by passing in a `containerProps` prop. Below shows the `Container` with different `maxWidth` and `backgroundColor` props. To see the button retain it's width at different ModalContent sizes see the [ModalContent size story](?path=/story/components-componentlibrary-modalcontent--size).
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-footer/README.mdx:102]:
```suggestion
The `ModalFooter` has an internal container that prevents the buttons from being large widths. You can override these props by passing in a `containerProps` prop. Below shows the `Container` with different `maxWidth` and `backgroundColor` props. To see the button retain it's width at different `ModalContent` sizes see the [ModalContent size story](?path=/story/components-componentlibrary-modalcontent--size).
```

File: ui/components/component-library/modal-footer/modal-footer.types.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+   */
+  submitButtonProps?: ButtonProps<'button'>;
+  /**
+   * The cancel button click event handler if this exists the cancel button will be displayed
+   */
+  onCancel?: () => void;
+  /**
+   * Additional props to pass to the cancel button
+   */
+  cancelButtonProps?: ButtonProps<'button'>;
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal-footer/modal-footer.types.ts:33]: Add `containerProps`

Comment (georgewrmarshall) [reply ui/components/component-library/modal-footer/modal-footer.types.ts:34]: Updated

File: ui/components/component-library/modal/modal.stories.tsx [hunk 8]
```diff
@@ -190,7 +186,7 @@ IsClosedOnEscapeKey.args = {
 };
 
 export const InitialFocusRef: StoryFn<typeof Modal> = (args) => {
-  const inputRef = React.useRef<HTMLDivElement>(null);
+  const inputRef = React.useRef<HTMLInputElement>(null);
   const [{ isOpen }, updateArgs] = useArgs();
   const handleOnClick = () => {
     updateArgs({ isOpen: true });
```
Comment (georgewrmarshall) [inline ui/components/component-library/modal/modal.stories.tsx:192]: Replacing placeholders with `ModalFooter` in `Modal` stories and docs

### Before


https://github.com/MetaMask/metamask-extension/assets/8112138/c5688ab2-4a5a-46ce-b4ee-2e10653020a1

### After


https://github.com/MetaMask/metamask-extension/assets/8112138/4cb548f4-51d5-418f-9c8e-82980ba0e691

File: ui/components/component-library/container/README.mdx (hunk 2 above)
Comment (georgewrmarshall) [reply ui/components/component-library/container/README.mdx]: Fixed in https://github.com/MetaMask/metamask-extension/pull/22432

**Review by garrettbear — APPROVED** [review]

**Review by david0xd — APPROVED** [review]

**Pull request merged**